### PR TITLE
mds: inode number range delegation to clients

### PR DESCRIPTION
**Pull request by jtlayton**
This is an initial set to have the MDS delegate ranges of the prealloc_inos set to the clients for use in async dirops. This is just a draft for now, and the kernel code does not exist yet.

The set starts with a few cleanup and bugfix patches that could probably be merged independently.

Old kernels will also barf when mounting an MDS with this set since they don't expect the extra field in an openc response. We'll also need to do something (possibly with feature bits) to work around that.

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/mds/mdstypes.cc [hunk 1]
```diff
@@ -532,7 +532,7 @@ void session_info_t::dump(Formatter *f) const
   f->open_array_section("prealloc_inos");
   for (const auto& [start, len] : prealloc_inos) {
```
Comment (ukernel) [inline src/mds/mdstypes.cc]: start is not range here.

Comment (jtlayton) [reply src/mds/mdstypes.cc]: Can you clarify what you mean? I'm not sure what you're saying here.

**Comment by jtlayton** [comment]
I've made the latest set conditional on the OCTOPUS cephfs feature bit, and fixed up the client decoder by defining a new (versioned) struct that we'll send instead of just the created inode number. gtests seem to be passing now.

**Comment by ukernel** [comment]
This code should works. But I forget the reason why we need to journal delegated_inos and save  delegated_inos in sessionmap object. I think we can simplify the code a lot if we just record  delegated_inos in memory

**Comment by jtlayton** [comment]
jenkins test make check

**Comment by jtlayton** [comment]
> This code should works. But I forget the reason why we need to journal delegated_inos and save delegated_inos in sessionmap object. I think we can simplify the code a lot if we just record delegated_inos in memory

You may be right. I think my thinking was:

Suppose I hold a buffered create request on the client (with a delegated inode number) and then the MDS crashes. It comes back and has no record of what inode numbers I have been delegated. The session comes back and a mkdir races in ahead of the buffered create, and MDS gives it the same inode number as the buffered create.

That said, I suppose we could just ensure that we flush out any buffered creates when reestablishing the session, and then that problem couldn't occur.

**Comment by ukernel** [comment]
On Tue, Nov 19, 2019 at 10:16 PM jtlayton <notifications@github.com> wrote:

> This code should works. But I forget the reason why we need to journal
> delegated_inos and save delegated_inos in sessionmap object. I think we can
> simplify the code a lot if we just record delegated_inos in memory
>
> You may be right. I think my thinking was:
>
> Suppose I hold a buffered create request on the client (with a delegated
> inode number) and then the MDS crashes. It comes back and has no record of
> what inode numbers I have been delegated. The session comes back and a
> mkdir races in ahead of the buffered create, and MDS gives it the same
> inode number as the buffered create.
>
> That said, I suppose we could just ensure that we flush out any buffered
> creates when reestablishing the session, and then that problem couldn't
> occur.
>
>
If mds crashes, client invalidates its delegated_inos. When mds recovers
and enter reconnect state, client flushes buffered creates (together with
unsafe requests). After MDS fully recovered, it can delegate inos to client
again.  If we do this, I think we can avoid saving delegated_inos to disk.

**Comment by jtlayton** [comment]
> If mds crashes, client invalidates its delegated_inos. When mds recovers and enter reconnect state, client flushes buffered creates (together with unsafe requests). After MDS fully recovered, it can delegate inos to client again. If we do this, I think we can avoid saving delegated_inos to disk.

Ok, that does sound reasonable. I'll rework this patchset along those lines and we'll just plan to ensure that async dirops are properly resent during recovery.

File: src/mds/mdstypes.cc (hunk 1 above)
Comment (ukernel) [reply src/mds/mdstypes.cc]: never mind.

File: src/mds/mdstypes.h [hunk 2]
```diff
@@ -1170,6 +1170,7 @@ struct session_info_t {
   entity_inst_t inst;
   std::map<ceph_tid_t,inodeno_t> completed_requests;
   interval_set<inodeno_t> prealloc_inos;   // preallocated, ready to use.
+  interval_set<inodeno_t> delegated_inos;  // hand these out to client
```
Comment (ukernel) [inline src/mds/mdstypes.h]: in-memory data structure shouldn't be defined in session_info_t.  delegated_inos should be defined in class Session.

Comment (ukernel) [reply src/mds/mdstypes.h]: If we don't erase delegated_inos from prealloc_inos. just let delegated_inos be subset of prealloc_inos.  _session_logged related changes can be avoid.

Comment (jtlayton) [reply src/mds/mdstypes.h]: I think I tried that at one point, but that means that we have to end up doing allocations (take_ino) from the middle of an interval_set, and the interval_set code doesn't really support doing that well. I'll try again to be sure, but I think this is actually simpler (sadly).

Comment (ukernel) [reply src/mds/mdstypes.h]: I saw warning removed by commit 'mds: don't warn when inodes are not allocated out sequentially' before.  For the warning case,  inode number is removed from middle of internal_set

**Comment by jtlayton** [comment]
Closing this PR for now, while I work out the new method.

**Pull request closed without merging**